daemon/containerd/image_exporter.go

```diff
@@ -41,33 +88,126 @@ func (i *ImageService) ExportImage(ctx context.Context, names []string, outStrea
-				return errors.Wrap(err, "failed to unpack image")
+				logger.WithError(err).Debug("failed to check if image is unpacked")
+				continue
 			}
+
+			if !unpacked {
+				err = platformImg.Unpack(ctx, i.snapshotter)
+
+				if err != nil {
+					return errdefs.System(err)
+				}
+			}
+			logger.WithField("alreadyUnpacked", unpacked).WithError(err).Debug("unpack")
 		}
+
+		fmt.Fprintf(progress, "Loaded image: %s\n", name)
 	}
 	return nil
 }
+
+// getBestDescriptorForExport returns a descriptor which only references content available locally.
+// The returned descriptor can be:
+// - The same index descriptor - if all content is available
+// - Platform specific manifest - if only one manifest from the whole index is available
+// - Reduced index descriptor - if not all, but more than one manifest is available
+func (i *ImageService) getBestDescriptorForExport(ctx context.Context, indexDesc ocispec.Descriptor) (ocispec.Descriptor, error) {
+	none := ocispec.Descriptor{}
+
+	if !containerdimages.IsIndexType(indexDesc.MediaType) {
+		err := fmt.Errorf("index/manifest-list descriptor expected, got: %s", indexDesc.MediaType)
+		return none, errdefs.InvalidParameter(err)
+	}
+	store := i.client.ContentStore()
+	children, err := containerdimages.Children(ctx, store, indexDesc)
+	if err != nil {
+		if cerrdefs.IsNotFound(err) {
+			return none, errdefs.NotFound(err)
+		}
+		return none, errdefs.System(err)
+	}
+
+	// Check which platform manifests have all their blobs available.
+	missingManifests := []ocispec.Descriptor{}
+	presentManifests := []ocispec.Descriptor{}
+	for _, mfst := range children {
+		if containerdimages.IsManifestType(mfst.MediaType) {
+			available, _, _, missing, err := containerdimages.Check(ctx, store, mfst, nil)
+			if err != nil {
+				missingManifests = append(missingManifests, mfst)
+				logrus.WithField("manifest", mfst.Digest).Warn("failed to check manifest's blob availability, won't export")
+				continue
+			}
+
+			if available && len(missing) == 0 {
+				presentManifests = append(presentManifests, mfst)
+				logrus.WithField("manifest", mfst.Digest).Debug("manifest content present, will export")
+			} else {
+				missingManifests = append(missingManifests, mfst)
+				logrus.WithFields(logrus.Fields{
+					"manifest": mfst.Digest,
+					"missing":  missing,
+				}).Debug("manifest is missing, won't export")
+			}
+		}
+	}
+
+	if len(missingManifests) == 0 {
+		// If we have the full image, just export the original index.
+		return indexDesc, nil
+
+	} else if len(presentManifests) == 1 {
+		// If only one platform is present, export that one manifest.
+		return presentManifests[0], nil
+	}
+
+	// Create a new index which contains only the manifests we have in store.
+	index := ocispec.Index{
+		Versioned: specs.Versioned{
+			SchemaVersion: 2,
+		},
+		MediaType: ocispec.MediaTypeImageIndex,
+		Manifests: presentManifests,
+	}
+
+	reducedIndexDesc, err := storeJson(ctx, store, index.MediaType, index, nil)
```

Comment:
I'm surprised that anything is written to the ContainerD content store as part of export. How come? Is it worth documenting the rationale in a code comment here?

What is the lifetime of this content blob? Could there be any issues with racing the garbage collector?

Reply:
It needs to be in the store for containerd's [archive.Export](https://github.com/containerd/containerd/blob/6ed24c88ed0530aa8d4483b13a217172365f61cc/images/archive/exporter.go#L190) to be able to read it.
We can't pass the manifest/index directly to the `Export`
It accepts `ocispec.Descriptor` of the manifests/indexes as an input, so we can't pass the Index directly.

Reply:
As for lifetime, thanks for pointing that one out. I forgot to add the lease to the context of these operations 🤦🏻 Fixed!
The lifetime of this blob is only this operation - it will be GCed after export is finished.

Reply:
We can wrap the store, though, and server it directly out of memory.